lib/BoundaryValueDiffEqAscher/src/utils.jl

```diff
@@ -0,0 +1,225 @@
+end
+function __get_bcresid_prototype(::TwoPointBVProblem, prob::BVProblem, u)
+    prototype = if prob.f.bcresid_prototype !== nothing
+        prob.f.bcresid_prototype.x
+    else
+        first(prob.f.bc)(u, prob.p), last(prob.f.bc)(u, prob.p)
+    end
+    return prototype, size.(prototype)
+end
+function __get_bcresid_prototype(::StandardBVProblem, prob::BVProblem, u)
+    prototype = prob.f.bcresid_prototype !== nothing ? prob.f.bcresid_prototype : zero(u)
+    return prototype, size(prototype)
+end
+
+# Custom pivot LU factorization and substitution for simple usage and
+# customized array types
+function __factorize!(a::Matrix{T}, ipvt::Vector) where {T}
+    t = 0.0
+    n = size(a, 1)
+    if n - 1 >= 1
+        for k in 1:(n - 1)
+            kp1 = k + 1
+            # find l = pivot index
+            l = argmax(abs.(a[k:end, k])) + k - 1
+            ipvt[k] = l
+            # zero pivot implies this column already triangularized
+            if a[l, k] !== 0
+                # interchange if necessary
+                if l !== k
+                    t = a[l, k]
+                    a[l, k] = a[k, k]
+                    a[k, k] = t
+                end
+                # compute  multipliers
+                t = -1.0 / a[k, k]
+                @views a[(k + 1):end, k] .= a[(k + 1):end, k] .* t
+                # row elimination with column indexing
+                for j in kp1:n
+                    t = a[l, j]
+                    if l !== k
+                        a[l, j] = a[k, j]
+                        a[k, j] = t
+                    end
+                    @views __muladd!(t, a[(k + 1):end, k], a[(k + 1):end, j])
+                end
+            end
+        end
+    end
+    ipvt[n] = n
+end
+
+function __substitute!(
+        a::Matrix{T1}, ipvt::Vector{I}, vb::Vector{Vector{T2}}) where {T1, T2, I}
+    n = size(a, 1)
+    b = reduce(vcat, vb)
+    if n - 1 >= 1
+        for k in 1:(n - 1)
+            l::I = ipvt[k]
+            t = b[l]
+            if l !== k
+                b[l] = b[k]
+                b[k] = t
+            end
+            @views __muladd!(t, a[(k + 1):end, k], b[(k + 1):end])
+        end
+    end
+    for k in n:-1:1
+        b[k] = b[k] / a[k, k]
+        @views __muladd!(-b[k], a[1:(k - 1), k], b[1:(k - 1)])
+    end
+    recursive_unflatten!(vb, b)
+end
+function __substitute!(
+        a::Matrix{T1}, ipvt::Vector{I}, b::AbstractVector{T2}) where {T1, T2 <: Real, I}
+    n = size(a, 1)
+    if n - 1 >= 1
+        for k in 1:(n - 1)
+            l::I = ipvt[k]
+            t = b[l]
+            if l !== k
+                b[l] = b[k]
+                b[k] = t
+            end
+            @views __muladd!(t, a[(k + 1):end, k], b[(k + 1):end])
+        end
+    end
+    for k in n:-1:1
+        b[k] = b[k] / a[k, k]
+        @views __muladd!(-b[k], a[1:(k - 1), k], b[1:(k - 1)])
+    end
+end
+
+@inline function __muladd!(a, x, y)
+    y .= muladd(a, x, y)
```

Comment:
```suggestion
    y .= muladd.(a, x, y)
```

It's going to allocate without the other `.`?

Reply:
Yeah, just fixed

Reply:
Yeah, just fixed